storage/tscache: Pick up andy-kimball/arenaskl fix

nvb opened the pull request:
Fixes #31624.
Fixes #35557.

This commit picks up andy-kimball/arenaskl#4.

I strongly suspect that the uint32 overflow fixed in that PR was the
cause of the two index out of bounds panics. See that commit for more
details.

The PR also fixes a bug in memory recylcling within the tscache. I confirmed
on adriatic that over 900 64MB arenas had been allocated since it was last
wiped.

nvb committed:
storage/tscache: fix skiplist arena re-use

This was incorrectly assuming that `Arena.Size` returned the full
capacity of the memory arena instead of the currently allocated
size. Since it is rare that the arena ever fills up without some
overflow, it was effectively disabling this optimization.

The size check turns out to not even be needed. I believe it was
added back when the page size was dynamic and hooked up to a cluster
setting. Since it isn't currently, we don't need the check.

Release note: None

nvb committed:
storage/tscache: Pick up andy-kimball/arenaskl fix

Fixes #31624.
Fixes #35557.

This commit picks up https://github.com/andy-kimball/arenaskl/pull/4.

I strongly suspect that the uint32 overflow fixed in that PR was the
cause of the two index out of bounds panics. See that commit for more
details.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35630)
<!-- Reviewable:end -->

ajwerner reviewed (approved):
This commit:
>See that commit for more details. 

That commit:
>The reason for this is subtle and has to do with failed allocations
slowly building up and eventually overflowing Arena.n. I'll explain
on the PR that vendors this fix.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> The other commit has reasonable commentary. You can let us in on further subtlety some other time. 

> Reviewed 1 of 1 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35630#-:-L_judhu1c-0k09EEOxk:b-y6ooef)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

nvb committed:
storage/tscache: return ErrArenaFull whenever a node cantInit

This change returns `ErrArenaFull` in all cases where `cantInit` is true
for a node in `ratchetValueSet`, instead of just when `cantInit` is true
and `ratchetValueSet` is attempting to initialize the node. We already
know any update to the node is going to fail, so there's no sense in
trying to update it again.

This should avoid a few memory copies and atomic loads for scans over nodes
that get stuck in a `cantInit` state in immutable `sklPages`.

Release note: None

nvanbenschoten commented:
The theory was that a full `sklPage` in `intervalSkl` with a large number of uninitialized nodes could end up resulting in a large number of failed attempts to allocate from the arena as [`scanTo`](https://github.com/cockroachdb/cockroach/blob/6a9c978b0b57a5405ab4c2bf1451b00c4398f2c2/pkg/storage/tscache/interval_skl.go#L1031) attempted to ratchet uninitialized nodes. This could drive up the arena size accounting without bound until it overflowed. This, in turn, could result in the `ErrArenaFull` check being bypassed and an allocation incorrectly appearing to succeed with an offset that would cause an out-of-bounds panic (after a corresponding underflow).

I added a third commit which should short-circuit the arena allocation attempts a little earlier for nodes that have already failed an arena allocation. This would have made the crash less frequent, but wouldn't have solved it entirely. However, this should improve performance in these cases where we already know an arena allocation is going to fail, which aren't uncommon for scans over immutable skl pages.

tbg reviewed (approved):
Good find.

nvanbenschoten commented:
bors r+
